pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,281 @@
+                
+                self._expr[field] = expr
+                scene_cfg[sname] = configs
+
+                for kc in configs:
+                    if kc == 'image_name' or kc == 'field_expr':
+                        pass
+                    elif kc == 'field':
+                        self.scenes[f's{i}/plots/p1/{kc}'] = field
+                    elif kc.startswith('render-'):
+                        self.scenes[f's{i}/renders/r1/{kc[7:]}'] = configs[kc]
+                    else:
+                        self.scenes[f's{i}/plots/p1/{kc}'] = configs[kc]
+            
+        return scene_cfg
+
+    def _config_pipeline(self, cfgsect):
+        pipel_cfg = {}
+        pl = ConduitNode()
+
+        i = -1 # pipeline counter
+        for k in self.cfg.items(cfgsect):
+            if k.startswith('pipeline-'):
+                i += 1
+                configs = self.cfg.getliteral(cfgsect, k)
+                configs = [configs] if isinstance(configs, dict) else configs
+                pname = k[9:]
+
+                pipel_cfg[pname] = configs
+                params = ConduitNode()
+                for (j, filter) in enumerate(configs):
+                    for kf in filter:
+                        if kf == 'type':
+                            pl[f'pl{i}/f{j}/{kf}'] = filter[kf]
+                        else:
+                            params[f'{kf}'] = filter[kf]
+
+                        if kf == 'field':
+                            try:
+                                _ = self._expr[kf]
+                            except KeyError:
+                                raise KeyError(f"Missing field '{kf}' used in pipeline")
+                pl[f'pl{i}/f{j}/params'] = params
+
+        self.pipelines = pl
+        return pipel_cfg
+
+    def _eval_exprs(self, intg):
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        # Iterate over each element type in the simulation
+        for i, (soln, (ename, eles)) in enumerate(zip(intg.soln, 
+                                                  self._ele_map.items())):
+            
+            dom_str = f'domain_{self.dom_0 + i}/'
+            # Subset and transpose the solution
+            soln = soln[...,:].swapaxes(0, 1)
+
+            mesh_op, soln_op = self.ops[ename]
+            # Interpolate the solution to the subdivided points
+            soln = soln_op @ soln
+
+            # Convert from conservative to primitive variables
+            psolns = self.elementscls.con_to_pri(soln, self.cfg)
+
+            # Prepare the substitutions dictionary
+            subs = dict(zip(pnames, psolns), t=intg.tcurr)
+            # subs |= dict(zip('xyz', plocs)) # TODO: add the ability for xyz in functions
+
+            # Prepare any required gradients
+            if self._gradpinfo:
+                # Compute the gradients
+                grad_soln = np.rollaxis(intg.grad_soln[i], 2)[..., :]
```

Comment:
May not be safe.

Reply:
in what way? This is lifted from the integrate plugin

Reply:
See recent issues we've had with the time averaging plugin (https://github.com/PyFR/PyFR/commit/4a6e0e593c6382537bcf573df72430229f8edb44)